README.md

```diff
@@ -6,7 +6,498 @@
+  the size of the transformed content. The `len` and `ensureRunway` APIs
+  make sure such cases are handled as optimally as possible.
+
+## Basic API usage
+
+The FastStreams API consists of ony few major object types:
+
+### `InputStream`
+
+An `InputStream` manages a particular input device. The library offers out
+of the box the following input stream types:
+
+* `fileInput`
+
+  For reading files through the familiar `fread` API from the C run-time.
+
+* `memFileInput`
+
+  For reading memory mapped files which provides best performance.
+
+* `unsafeMemoryInput`
+
+  For handling strings, sequences and openarrays as an input stream. <br />
+  You are responsible for ensuring that the backing buffer won't be invalidated
+  while the stream is being used.
+
+* `memoryInput`
+
+  Primarily used to consume the contents written to a previously populated
+  output stream, but it can also be used to consume the contents of strings
+  and sequences in a memory-safe way (by creating a copy).
+
+* `pipeInput` (async)
+
+  For arbitrary conmmunication between a produced and a consumer.
+
+* `chronosInput` (async)
+
+  Enabled by importing `faststreams/chronos_adapters`. <br />
+  It can represent any Chronos `Transport` as an input stream.
+
+* `asyncSocketInput` (async)
+
+  Enabled by importing `faststreams/std_adapters`. <br />
+  Allows using Nim's standard library `AsyncSocket` type as an input stream.
+
+You can extend the library with new `InputStream` types without modifying it.
+Please see the inline code documentation of `InputStreamVTable` for more details.
+
+All of the above APIs are possible constructors for creating an `InputStream`.
+The stream instances will manage their resources through destructors, but you
+might want to `close` them explicitly in async context or when you need to
+handle the possible errors from the closing operation.
+
+Here is an example usage:
+
+```nim
+import
+  faststreams/inputs
+
+var
+  jsonString = "[1, 2, 3]"
+  jsonNodes = parseJson(unsafeMemoryInput(jsonString))
+  moreNodes = parseJson(fileInput("data.json"))
+```
+
+The example above assumes we might have a `parseJson` function accepting an
+`InputStream`. Here how this function could be defined:
+
+```nim
+import
+  faststreams/inputs
+
+proc scanString(stream: InputStream): JsonToken {.fsMultiSync.} =
+  result = newStringToken()
+
+  advance stream # skip the opening quote
+
+  while stream.readable:
+    let nextChar = stream.read.char
+    case nextChar
+    of '\'':
+      if stream.readable:
+        let escaped = stream.read.char
+        case escaped
+        of 'n': result.add '\n'
+        of 't': result.add '\t'
+        else: result.add escaped
+      else:
+        error(UnexpectedEndOfFile)
+    of '"'
+      return
+    else:
+      result.add nextChar
+
+  error(UnexpectedEndOfFile)
+
+proc nextToken(stream: InputStream): JsonToken {.fsMultiSync.} =
+  while stream.readable:
```

Comment:
if this is the only async point, why not make it `while await stream.readable`? for sync streams, this would be a noop, for async streams it would do the future dance (if necessary).

it remains the case that nim has no safeguards against concurrent modification - teaching developers yet another hidden out-of-order execution keyword is a high cost in terms of bugs - in terms of syntactic overhead, this is almost nothing, but reminds the developer that _this_ is where the synchronization happens, and not in the `read` below, in a way that's established across libraries and frameworks, nim and not

Reply:
notably, `readable` does not necessarily have to be future-based - the new `await` (perhaps better call it `fsawait` to not overload `await` until it's clear from common usage that it's fine) could be a noop for sync streams and not require `Future` return types from whatever it calls - if it returns a future, it would do the future dance, if it returns `Opt[future]` or some such it would do the "do I need to alloc-a-future"-thing etc

Reply:
It's possible to implement this and I might agree with it if it's what it takes to get wider adoption of the API, but I consider the argument a bit moot, because in the README I've highlighted that the FastStreams API is mostly useful for implementing:


* Parsers for data formats and protocols employing formal grammars
* Block ciphers
* Compressors and decompressors
* Stream multiplexers

Think about your concerns in this context. What is the mutable state in a block compressor or a Json parser? These are stream processing algorithms that are very much focused on their inputs and outputs and that benefit significantly from the "just like sync code" mentally of the FastStreams API.

Reply:
well - the output goes somewhere - and as soon as you get into use cases where performance actually matters, as opposed to toy examples and config files, people will want to optimize where that is - for example a state variable (say.. a beacon state) that might get used or mutated from outside because nim doesn't track ownership. 

This means in particular, that for an extended period of time, partially updated instances will live in memory and be accessible "outside" - clever optimizations like RVO even mutate instances in a totally unexpected way - even the most experienced compiler developers get this wrong though they intimately know the feature - so you can't even blame lazy devs that didn't read the manual (https://github.com/nim-lang/Nim/issues/14126).

On a network (where async is predominatly used), data trickles slowly, with latency, errors, disconnects and all kinds of erratic behaviours - to write something that approaches correct and performant behaviour, you need to adapt your code to where the reading and waiting happens, and for that, you need to know and see it. A readme in a far-off repository doesn't count, because that readme is not on your mind when you're trying to solve a problem - you're busy with something else.

In short, you can't responsibly hide this stuff, it only works when it doesn't matter - on the toy examples that devs will try and call "testing".

Reply:
The output will usually go at the end of a "Pipeline". Here is one example:

https://github.com/status-im/nim-beacon-chain/blob/b41f4347731585f66630a1a02e3f15326acaf4a8/beacon_chain/faststreams_backend.nim#L90-L91

Such pipelines are usually awaited.

It also occurred to me that there is not much evidence that a visible `await` prevents the problem. Writing `await` is pretty routine and people don't usually stop to think about these issues. I'll agree that someone who is extremely knowledgable about async issues, but not familiar with FastStreams may benefit from a visible `fsAwait` during code review (well, she'll have to know what that is).

Reply:
anybody who is not knowledgeable about `async` issues will contribute broken code to the world with this API - it's inviting the developer to not understand the issues while not doing the right things to solve them.

writing `await` gives whoever is reading the code a hint about what might be wrong, and where - much like a function signature, it's a marker that shows where specific issues need to be considered - this is helpful not only during development but also later, during debugging. it also opens up the mind to the right google terms to search for, which gives access to a wide range of resources across languages that explain these issues.